drivers/video/video_emul_imager.c

```diff
@@ -0,0 +1,497 @@
+
+	/* Fill the first row of the emulated framebuffer */
+	switch (fmt->pixelformat) {
+	case VIDEO_PIX_FMT_YUYV:
+		for (size_t i = 0; i < fmt->width; i++) {
+			y = pattern_8bars_yuv[i * 8 / fmt->width][0];
+			uv = pattern_8bars_yuv[i * 8 / fmt->width][1 + i % 2];
+			fb16[i] = sys_cpu_to_be16(y << 8 | uv << 0);
+		}
+		break;
+	case VIDEO_PIX_FMT_RGB565:
+		for (size_t i = 0; i < fmt->width; i++) {
+			r = pattern_8bars_rgb[i * 8 / fmt->width][0] >> 3;
+			g = pattern_8bars_rgb[i * 8 / fmt->width][1] >> 2;
+			b = pattern_8bars_rgb[i * 8 / fmt->width][2] >> 3;
+			fb16[i] = sys_cpu_to_le16(r << 11 | g << 5 | b << 0);
+		}
+		break;
+	default:
+		LOG_WRN("Unsupported pixel format %x, supported: %x, %x", fmt->pixelformat,
+			VIDEO_PIX_FMT_YUYV, VIDEO_PIX_FMT_RGB565);
+		memset(fb16, 0, fmt->pitch);
+	}
+
+	/* Duplicate the first row over the whole frame */
+	for (size_t i = 1; i < fmt->height; i++) {
+		memcpy(data->framebuffer + fmt->pitch * i, data->framebuffer, fmt->pitch);
+	}
+}
+
+static int emul_imager_set_fmt(const struct device *const dev, enum video_endpoint_id ep,
+			       struct video_format *fmt)
+{
+	struct emul_imager_data *data = dev->data;
+	size_t fmt_id;
+	int ret;
+
+	if (ep != VIDEO_EP_OUT && ep != VIDEO_EP_ALL) {
+		return -EINVAL;
+	}
+
+	if (fmt->pitch * fmt->height > CONFIG_VIDEO_EMUL_IMAGER_FRAMEBUFFER_SIZE) {
+		LOG_ERR("%s has %u bytes of memory, unable to support %x %ux%u (%u bytes)",
+			dev->name, CONFIG_VIDEO_EMUL_IMAGER_FRAMEBUFFER_SIZE, fmt->pixelformat,
+			fmt->width, fmt->height, fmt->pitch * fmt->height);
+		return -ENOMEM;
+	}
+
+	if (memcmp(&data->fmt, fmt, sizeof(data->fmt)) == 0) {
+		return 0;
+	}
+
+	ret = video_format_caps_index(fmts, fmt, &fmt_id);
+	if (ret < 0) {
+		LOG_ERR("Format %x %ux%u not found for %s", fmt->pixelformat, fmt->width,
+			fmt->height, dev->name);
+		return ret;
+	}
+
+	ret = emul_imager_set_mode(dev, &emul_imager_modes[fmt_id][0]);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Change the image pattern on the framebuffer */
+	emul_imager_fill_framebuffer(dev, fmt);
+
+	data->fmt_id = fmt_id;
+	data->fmt = *fmt;
+	return 0;
+}
+
+static int emul_imager_get_fmt(const struct device *dev, enum video_endpoint_id ep,
+			       struct video_format *fmt)
+{
+	struct emul_imager_data *data = dev->data;
+
+	if (ep != VIDEO_EP_OUT && ep != VIDEO_EP_ALL) {
+		return -EINVAL;
+	}
+
+	*fmt = data->fmt;
+	return 0;
+}
+
+static int emul_imager_get_caps(const struct device *dev, enum video_endpoint_id ep,
+				struct video_caps *caps)
+{
+	if (ep != VIDEO_EP_OUT && ep != VIDEO_EP_ALL) {
+		return -EINVAL;
+	}
+
+	caps->format_caps = fmts;
+	return 0;
+}
+
+static int emul_imager_stream_start(const struct device *dev)
+{
+	return emul_imager_write_reg(dev, EMUL_IMAGER_REG_CTRL, 1);
+}
+
+static int emul_imager_stream_stop(const struct device *dev)
+{
+	return emul_imager_write_reg(dev, EMUL_IMAGER_REG_CTRL, 0);
+}
+
+static const struct video_driver_api emul_imager_driver_api = {
```

Comment:
Please use the `DEVICE_API` wrapper.

```suggestion
static DEVICE_API(video, emul_imager_driver_api) = {
```

Reply:
Yes thank you, done.
This will be helpful for checking device types in the video shell based on this PR!